sqlglot/optimizer/rewrite_derived_tables_as_ctes.py

```diff
@@ -0,0 +1,40 @@
+from sqlglot import expressions as exp
+from sqlglot.helper import find_new_name
+from sqlglot.optimizer.scope import traverse_scope
+
+
+def rewrite_derived_tables_as_ctes(expression):
+    """
+    Rewrite derived tables as common table expressions.
+
+    Example:
+        >>> import sqlglot
+        >>> expression = sqlglot.parse_one("SELECT a FROM (SELECT * FROM x) AS y")
+        >>> rewrite_derived_tables_as_ctes(expression).sql()
+        'WITH y AS (SELECT * FROM x) SELECT a FROM y'
+
+    Args:
+        expression (sqlglot.Expression): expression to optimize
+    Returns:
+        sqlglot.Expression: optimized expression
+    """
+    expression = expression.copy()
```

Comment:
optimizer methods shouldn't have copies, the implicit standard is for the caller to call copy

Reply:
good call